File: backend/review-service/src/main/java/com/livinglink/reviewservice/service/ReviewService.java
package com.livinglink.reviewservice.service;

import com.livinglink.reviewservice.client.VisitClient;
import com.livinglink.reviewservice.dto.ReviewRequestDto;
import com.livinglink.reviewservice.dto.ReviewSummaryDto;
import com.livinglink.reviewservice.dto.VisitDto;
import com.livinglink.reviewservice.entity.Review;
import com.livinglink.reviewservice.exception.BadRequestException;
import com.livinglink.reviewservice.exception.DuplicateResourceException;
import com.livinglink.reviewservice.exception.ResourceNotFoundException;
import com.livinglink.reviewservice.exception.UnauthorizedException;
import com.livinglink.reviewservice.repository.ReviewRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
public class ReviewService {

    private final ReviewRepository reviewRepository;
    private final VisitClient visitClient;

    public ReviewService(ReviewRepository reviewRepository, VisitClient visitClient) {
        this.reviewRepository = reviewRepository;
        this.visitClient = visitClient;
    }

    public Review createReview(Long reviewerId, ReviewRequestDto request) {
        if ("ACCOMMODATION".equals(request.getReviewType())) {
            if (request.getListingId() == null) {
                throw new BadRequestException("listingId is required for ACCOMMODATION review");
            }
            if (reviewRepository.existsByReviewerIdAndListingIdAndReviewType(reviewerId, request.getListingId(), "ACCOMMODATION")) {
                throw new DuplicateResourceException("You have already reviewed this listing");
            }
            verifyAccommodationRelationship(request.getListingId());
        } else if ("ROOMMATE".equals(request.getReviewType())) {
            if (request.getReviewedUserId() == null) {
                throw new BadRequestException("reviewedUserId is required for ROOMMATE review");
            }
            if (reviewerId.equals(request.getReviewedUserId())) {
                throw new BadRequestException("You cannot review yourself");
            }
            if (reviewRepository.existsByReviewerIdAndReviewedUserIdAndReviewType(reviewerId, request.getReviewedUserId(), "ROOMMATE")) {
                throw new DuplicateResourceException("You have already reviewed this user");
            }
            verifyRoommateRelationship(request.getReviewedUserId());
        } else {
            throw new BadRequestException("Invalid reviewType. Must be ACCOMMODATION or ROOMMATE");
        }

        Review review = new Review();
        review.setReviewerId(reviewerId);
        review.setReviewedUserId(request.getReviewedUserId());
        review.setListingId(request.getListingId());
        review.setRating(request.getRating());
        review.setComment(request.getComment());
        review.setReviewType(request.getReviewType());
        review.setCreatedAt(LocalDateTime.now());
        review.setUpdatedAt(LocalDateTime.now());

        return reviewRepository.save(review);
    }

    private void verifyAccommodationRelationship(Long listingId) {
        // Fetch visits where current user is requester
        List<VisitDto> myRequests = visitClient.getMyVisitRequests();
        
        boolean hasAcceptedVisit = myRequests.stream()
                .anyMatch(v -> v.getListingId().equals(listingId) && "ACCEPTED".equals(v.getStatus()));

        if (!hasAcceptedVisit) {
            throw new UnauthorizedException("You can only review accommodations you have an ACCEPTED visit for.");
        }
    }

    private void verifyRoommateRelationship(Long reviewedUserId) {
        List<VisitDto> myRequests = visitClient.getMyVisitRequests();
        List<VisitDto> myOwnerVisits = visitClient.getVisitsForMyListings();

        boolean sharedVisit = myRequests.stream()
                .anyMatch(v -> v.getOwnerId().equals(reviewedUserId) && "ACCEPTED".equals(v.getStatus()))
                ||
                myOwnerVisits.stream()
                .anyMatch(v -> v.getRequesterId().equals(reviewedUserId) && "ACCEPTED".equals(v.getStatus()));

        if (!sharedVisit) {
            throw new UnauthorizedException("You can only review users you have shared an ACCEPTED visit with.");
        }
    }

    public List<Review> getReviewsForListing(Long listingId) {
        return reviewRepository.findByListingId(listingId);
    }

    public List<Review> getReviewsForUser(Long userId) {
        return reviewRepository.findByReviewedUserId(userId);
    }

    public ReviewSummaryDto getListingReviewSummary(Long listingId) {
        List<Review> reviews = getReviewsForListing(listingId);
        return calculateSummary(listingId, reviews);
    }

    public ReviewSummaryDto getUserReviewSummary(Long userId) {
        List<Review> reviews = getReviewsForUser(userId);
        return calculateSummary(userId, reviews);
    }

    private ReviewSummaryDto calculateSummary(Long targetId, List<Review> reviews) {
        if (reviews.isEmpty()) {
            return new ReviewSummaryDto(targetId, 0.0, 0);
        }
        
        double avg = reviews.stream()
                .mapToInt(Review::getRating)
                .average()
                .orElse(0.0);
                
        // Round to 1 decimal place
        avg = Math.round(avg * 10.0) / 10.0;
        
        return new ReviewSummaryDto(targetId, avg, reviews.size());
    }

    public Review updateReview(Long reviewId, Long reviewerId, ReviewRequestDto request) {
        Review review = reviewRepository.findById(reviewId)
                .orElseThrow(() -> new ResourceNotFoundException("Review not found"));

        if (!review.getReviewerId().equals(reviewerId)) {
            throw new UnauthorizedException("You can only update your own reviews");
        }

        review.setRating(request.getRating());
        review.setComment(request.getComment());
        review.setUpdatedAt(LocalDateTime.now());

        return reviewRepository.save(review);
    }

    public void deleteReview(Long reviewId, Long reviewerId) {
        Review review = reviewRepository.findById(reviewId)
                .orElseThrow(() -> new ResourceNotFoundException("Review not found"));

        if (!review.getReviewerId().equals(reviewerId)) {
            throw new UnauthorizedException("You can only delete your own reviews");
        }

        reviewRepository.delete(review);
    }
}

File: frontend/livinglink-client/src/services/reviewService.js
import api from "./api";

/**
 * Review Service
 * Connects to review-service via API Gateway at /api/reviews
 */

/**
 * Create a new review
 * POST /api/reviews
 *
 * Payload format:
 * {
 *   listingId?: number,
 *   reviewedUserId?: number,
 *   rating: number (1-5),
 *   comment: string (5-1000 chars),
 *   reviewType: "ACCOMMODATION" | "ROOMMATE"
 * }
 */
export const createReview = async (reviewData) => {
  const payload = {
    rating: Number(reviewData.rating),
    comment: reviewData.comment?.trim(),
    reviewType: reviewData.reviewType || (reviewData.accommodationId || reviewData.listingId ? "ACCOMMODATION" : "ROOMMATE"),
  };

  if (reviewData.accommodationId || reviewData.listingId) {
    payload.listingId = Number(reviewData.accommodationId || reviewData.listingId);
  }

  if (reviewData.reviewedUserId || reviewData.userId) {
    payload.reviewedUserId = Number(reviewData.reviewedUserId || reviewData.userId);
  }

  const response = await api.post("/api/reviews", payload);
  return response.data;
};

/**
 * Get all reviews for a listing/accommodation
 * GET /api/reviews/listing/{listingId}
 */
export const getAccommodationReviews = async (accommodationId) => {
  const response = await api.get(`/api/reviews/listing/${accommodationId}`);
  return Array.isArray(response.data) ? response.data : [];
};

/**
 * Get review summary (average rating & count) for a listing
 * GET /api/reviews/listing/{listingId}/summary
 */
export const getListingReviewSummary = async (accommodationId) => {
  const response = await api.get(`/api/reviews/listing/${accommodationId}/summary`);
  return response.data || { targetId: Number(accommodationId), averageRating: 0, reviewCount: 0 };
};

/**
 * Get all reviews for a user (as a roommate)
 * GET /api/reviews/user/{userId}
 */
export const getRoommateReviews = async (userId) => {
  const response = await api.get(`/api/reviews/user/${userId}`);
  return response.data;
};

/**
 * Get review summary for a user
 * GET /api/reviews/user/{userId}/summary
 */
export const getUserReviewSummary = async (userId) => {
  const response = await api.get(`/api/reviews/user/${userId}/summary`);
  return response.data;
};

/**
 * Update an existing review
 * PUT /api/reviews/{id}
 */
export const updateReview = async (id, reviewData) => {
  const response = await api.put(`/api/reviews/${id}`, reviewData);
  return response.data;
};

/**
 * Delete a review by ID
 * DELETE /api/reviews/{id}
 */
export const deleteReview = async (id) => {
  const response = await api.delete(`/api/reviews/${id}`);
  return response.data;
};

const reviewService = {
  createReview,
  getAccommodationReviews,
  getListingReviewSummary,
  getRoommateReviews,
  getUserReviewSummary,
  updateReview,
  deleteReview,
};

export default reviewService;
